<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Broad Institute" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>SRJointCallGVCFsWithGenomicsDB - Broad Institute Long Read Pipelines</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "SRJointCallGVCFsWithGenomicsDB";
        var mkdocs_page_input_path = "workflows/SRJointCallGVCFsWithGenomicsDB.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> Broad Institute Long Read Pipelines
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Development guide</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/">Development Guide Overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/docker_style_guide/">Docker Style Guide</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/repo_structure/">Repository Structure</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/wdl_style_guide/">WDL Style Guide</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">Sp malaria</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../pipeline_documentation/sp_malaria/workspace_markdown_doc/">_P. falciparum_ Short Read Whole Genome Workspace</a>
                </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/AlignReads/">AlignReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/BroadOnPremMalariaPipelineTasks/">BroadOnPremMalariaPipelineTasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Clair/">Clair</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/FastQC/">FastQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Finalize/">Finalize</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Fingerprinting/">Fingerprinting</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/FunctionalAnnotation/">FunctionalAnnotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/GeneralUtils/">GeneralUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Hail/">Hail</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MASSeq/">MASSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MalariaReports/">MalariaReports</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MalariaReportsSnapshots/">MalariaReportsSnapshots</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Medaka/">Medaka</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/NanoPlot/">NanoPlot</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ONTPepper/">ONTPepper</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ONTUtils/">ONTUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/PBUtils/">PBUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Postprocessing_Tasks/">Postprocessing_Tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Preprocessing_Tasks/">Preprocessing_Tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/QCAssessment/">QCAssessment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Quast/">Quast</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SGKit/">SGKit</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SRJointGenotyping/">SRJointGenotyping</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SRUtils/">SRUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/StringTie2/">StringTie2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Utils/">Utils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/VariantUtils/">VariantUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cnv_common_tasks/">cnv_common_tasks</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../AlignAndCheckFingerprintCCS/">AlignAndCheckFingerprintCCS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AlignedMetrics/">AlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AnnotateTranscriptomeWithGuide/">AnnotateTranscriptomeWithGuide</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AssignChildLongReads/">AssignChildLongReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BackupWorkspace/">BackupWorkspace</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BenchmarkVCFs/">BenchmarkVCFs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BroadOnPremMalariaPipeline_1_Alignment_and_Variant_Calling/">BroadOnPremMalariaPipeline_1_Alignment_and_Variant_Calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BroadOnPremMalariaPipeline_2_JointVariantCalling/">BroadOnPremMalariaPipeline_2_JointVariantCalling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CCSPepper/">CCSPepper</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallAssemblyVariants/">CallAssemblyVariants</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsIllumina/">CallVariantsIllumina</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsONT/">CallVariantsONT</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsPBCCS/">CallVariantsPBCCS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Canu/">Canu</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CleanupIntermediate/">CleanupIntermediate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectPacBioAlignedMetrics/">CollectPacBioAlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectParentsKmerStats/">CollectParentsKmerStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectSMRTCellUnalignedMetrics/">CollectSMRTCellUnalignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CompareVcfBenchmarks/">CompareVcfBenchmarks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ConvertToHailMT/">ConvertToHailMT</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ConvertToZarrStore/">ConvertToZarrStore</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DeepTrio/">DeepTrio</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DeepVariant/">DeepVariant</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromFTP/">DownloadFromFTP</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromSRA/">DownloadFromSRA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromWeb/">DownloadFromWeb</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExpandedDrugResistanceMarkerAggregation/">ExpandedDrugResistanceMarkerAggregation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExpandedDrugResistanceMarkerExtraction/">ExpandedDrugResistanceMarkerExtraction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExtractRegionsFromBam/">ExtractRegionsFromBam</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExtractSnpBarcodeFromVcfFile/">ExtractSnpBarcodeFromVcfFile</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../FPCheckAoU/">FPCheckAoU</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../FindBamIdentity/">FindBamIdentity</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Flye/">Flye</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../GLNexus/">GLNexus</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Guppy/">Guppy</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../HaplotypeCaller/">HaplotypeCaller</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Hifiasm/">Hifiasm</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LRCNVs/">LRCNVs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LRConvertBCF/">LRConvertBCF</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LRJointCallGVCFs/">LRJointCallGVCFs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Longbow/">Longbow</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MalariaReportsWorkflow/">MalariaReportsWorkflow</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTAssembleWithCanu/">ONTAssembleWithCanu</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTAssembleWithFlye/">ONTAssembleWithFlye</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTBasecall/">ONTBasecall</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTFlowcell/">ONTFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTFlowcellFromMultipleBasecalls/">ONTFlowcellFromMultipleBasecalls</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTMethylation/">ONTMethylation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTProcessBasecall/">ONTProcessBasecall</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTWholeGenome/">ONTWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBAssembleWithHifiasm/">PBAssembleWithHifiasm</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBCCSIsoSeq/">PBCCSIsoSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBCCSWholeGenome/">PBCCSWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBFlowcell/">PBFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBMASIsoSeqDemultiplex/">PBMASIsoSeqDemultiplex</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBMASIsoSeqQuantify/">PBMASIsoSeqQuantify</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBSV/">PBSV</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PanelProcessMalariaBarcodesForRh/">PanelProcessMalariaBarcodesForRh</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Pf_Niare_HaplotypeCaller/">Pf_Niare_HaplotypeCaller</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumDrugResistanceSummary/">PfalciparumDrugResistanceSummary</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumHrp2Hrp3Status/">PfalciparumHrp2Hrp3Status</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumPolygenomicityBarcodeEstimate/">PfalciparumPolygenomicityBarcodeEstimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumTypeDrugResistanceMarkers/">PfalciparumTypeDrugResistanceMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ProcessMalariaBarcodesDemo/">ProcessMalariaBarcodesDemo</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../QCAssessmentDepthBased/">QCAssessmentDepthBased</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../RemoveSingleOrganismContamination/">RemoveSingleOrganismContamination</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRBamToFq/">SRBamToFq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRFlowcell/">SRFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRIndexBam/">SRIndexBam</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">SRJointCallGVCFsWithGenomicsDB</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#srjointcallgvcfswithgenomicsdb_1">SRJointCallGVCFsWithGenomicsDB</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#dot-diagram">Dot Diagram</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS/">SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VQSR/">SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VQSR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome/">SRWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome_Pf_Niare_VETS/">SRWholeGenome_Pf_Niare_VETS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome_Pf_Niare_VQSR/">SRWholeGenome_Pf_Niare_VQSR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SampleLevelAlignedMetrics/">SampleLevelAlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Sniffles2/">Sniffles2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SvQCPlots/">SvQCPlots</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TrainCnnFilters/">TrainCnnFilters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TrioBinChildLongReads/">TrioBinChildLongReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../VerifyFingerprint/">VerifyFingerprint</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">Broad Institute Long Read Pipelines</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Workflows &raquo;</li>
      <li>SRJointCallGVCFsWithGenomicsDB</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/broadinstitute/long-read-pipelines/edit/master/docs/workflows/SRJointCallGVCFsWithGenomicsDB.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="srjointcallgvcfswithgenomicsdb">SRJointCallGVCFsWithGenomicsDB</h1>
<h2 id="srjointcallgvcfswithgenomicsdb_1">SRJointCallGVCFsWithGenomicsDB</h2>
<dl>
<dt>author</dt>
<dd>Jonn Smith</dd>
<dt>description</dt>
<dd>A workflow that performs joint calling on single-sample gVCFs from GATK4 HaplotypeCaller using GenomicsDB.</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>gvcf_indices</code> (Array[File], <strong>required</strong>): Array of gvcf index files for <code>gvcfs</code>.  Order should correspond to that in <code>gvcfs</code>.</li>
<li><code>gvcfs</code> (Array[File], <strong>required</strong>): Array of GVCF files to use as inputs for joint calling.</li>
<li><code>indel_is_calibration</code> (Array[Boolean], <strong>required</strong>): Array of booleans indicating which files in <code>indel_known_reference_variants</code> should be used as calibration sets.  True -&gt;calibration set.  False -&gt; NOT a calibration set.</li>
<li><code>indel_is_training</code> (Array[Boolean], <strong>required</strong>): Array of booleans indicating which files in <code>indel_known_reference_variants</code> should be used as training sets.  True -&gt; training set.  False -&gt; NOT a training set.</li>
<li><code>indel_known_reference_variants</code> (Array[File], <strong>required</strong>): Array of VCF files to use as input reference variants for INDELs.  Each can be designated as either calibration or training using <code>indel_is_training</code> and <code>indel_is_calibration</code>.</li>
<li><code>indel_known_reference_variants_identifier</code> (Array[File], <strong>required</strong>): Array of names to give to the VCF files given in <code>indel_known_reference_variants</code>.  Order should correspond to that in <code>indel_known_reference_variants</code>.</li>
<li><code>indel_known_reference_variants_index</code> (Array[File], <strong>required</strong>): Array of VCF index files for <code>indel_known_reference_variants</code>.  Order should correspond to that in <code>indel_known_reference_variants</code>.</li>
<li><code>prefix</code> (String, <strong>required</strong>): Prefix to use for output files.</li>
<li><code>ref_map_file</code> (File, <strong>required</strong>): Reference map file indicating reference sequence and auxillary file locations</li>
<li><code>snp_is_calibration</code> (Array[Boolean], <strong>required</strong>): Array of booleans indicating which files in <code>snp_known_reference_variants</code> should be used as calibration sets.  True -&gt;calibration set.  False -&gt; NOT a calibration set.</li>
<li><code>snp_is_training</code> (Array[Boolean], <strong>required</strong>): Array of booleans indicating which files in <code>snp_known_reference_variants</code> should be used as training sets.  True -&gt; training set.  False -&gt; NOT a training set.</li>
<li><code>snp_known_reference_variants</code> (Array[File], <strong>required</strong>): Array of VCF files to use as input reference variants for SNPs.  Each can be designated as either calibration or training using <code>snp_is_training</code> and <code>snp_is_calibration</code>.</li>
<li><code>snp_known_reference_variants_identifier</code> (Array[File], <strong>required</strong>): Array of names to give to the VCF files given in <code>snp_known_reference_variants</code>.  Order should correspond to that in <code>snp_known_reference_variants</code>.</li>
<li><code>snp_known_reference_variants_index</code> (Array[File], <strong>required</strong>): Array of VCF index files for <code>snp_known_reference_variants</code>.  Order should correspond to that in <code>snp_known_reference_variants</code>.</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>annotation_bed_file_annotation_names</code> (Array[String]?): Array of names/FILTER column entries to use for each given file in <code>annotation_bed_files</code>.  Order should correspond to <code>annotation_bed_files</code>.</li>
<li><code>annotation_bed_file_indexes</code> (Array[File]?): Array of bed indexes for <code>annotation_bed_files</code>.  Order should correspond to <code>annotation_bed_files</code>.</li>
<li><code>annotation_bed_files</code> (Array[File]?): Array of bed files to use to FILTER/annotate variants in the output file.  Annotations will be placed in the FILTER column, effectively filtering variants that overlap these regions.</li>
<li><code>background_sample_gvcf_indices</code> (Array[Array[File]]?): Array of GVCF index files for <code>background_sample_gvcfs</code>.  Order should correspond to that in <code>background_sample_gvcfs</code>.</li>
<li><code>background_sample_gvcfs</code> (Array[Array[File]]?): Array of GVCFs to use as background samples for joint calling.</li>
<li><code>gcs_out_root_dir</code> (String?): GCS Bucket into which to finalize outputs.  If no bucket is given, outputs will not be finalized and instead will remain in their native execution location.</li>
<li><code>interval_list</code> (File?)</li>
<li><code>snpeff_db</code> (File?)</li>
<li><code>snpeff_db_identifier</code> (String?)</li>
<li><code>AnnotateVcfRegions.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ConvertToZarr.ref_fai</code> (String?)</li>
<li><code>ConvertToZarr.ref_fasta</code> (String?)</li>
<li><code>ConvertToZarr.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>CreateHailMatrixTable.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>CreateIntervalListFileFromIntervalInfo.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>CreateSampleNameMap.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ExtractIndelVariantAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ExtractIntervalNamesFromIntervalOrBamFile.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ExtractSnpVariantAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeGenomicsDB.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeHailMatrixTable.name</code> (String?)</li>
<li><code>FinalizeHailMatrixTable.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelExtractedAnnotations.name</code> (String?)</li>
<li><code>FinalizeIndelExtractedAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelExtractedSitesOnlyVcf.name</code> (String?)</li>
<li><code>FinalizeIndelExtractedSitesOnlyVcf.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelExtractedSitesOnlyVcfIndex.name</code> (String?)</li>
<li><code>FinalizeIndelExtractedSitesOnlyVcfIndex.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelExtractedUnlabeledAnnotations.name</code> (String?)</li>
<li><code>FinalizeIndelExtractedUnlabeledAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsCalibrationSetScores.name</code> (String?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsCalibrationSetScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsNegativeModelScorer.name</code> (String?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsNegativeModelScorer.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsPositiveModelScorer.name</code> (String?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsPositiveModelScorer.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsTrainingScores.name</code> (String?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsTrainingScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsUnlabeledPositiveModelScores.name</code> (String?)</li>
<li><code>FinalizeIndelTrainVariantAnnotationsUnlabeledPositiveModelScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsAnnotationsHdf5.name</code> (String?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsAnnotationsHdf5.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoredVcf.name</code> (String?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoredVcf.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoredVcfIndex.name</code> (String?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoredVcfIndex.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoresHdf5.name</code> (String?)</li>
<li><code>FinalizeScoreIndelVariantAnnotationsScoresHdf5.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsAnnotationsHdf5.name</code> (String?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsAnnotationsHdf5.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoredVcf.name</code> (String?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoredVcf.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoredVcfIndex.name</code> (String?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoredVcfIndex.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoresHdf5.name</code> (String?)</li>
<li><code>FinalizeScoreSnpVariantAnnotationsScoresHdf5.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpEffGenes.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpEffSummary.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpExtractedAnnotations.name</code> (String?)</li>
<li><code>FinalizeSnpExtractedAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpExtractedSitesOnlyVcf.name</code> (String?)</li>
<li><code>FinalizeSnpExtractedSitesOnlyVcf.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpExtractedSitesOnlyVcfIndex.name</code> (String?)</li>
<li><code>FinalizeSnpExtractedSitesOnlyVcfIndex.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpExtractedUnlabeledAnnotations.name</code> (String?)</li>
<li><code>FinalizeSnpExtractedUnlabeledAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsCalibrationSetScores.name</code> (String?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsCalibrationSetScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsNegativeModelScorer.name</code> (String?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsNegativeModelScorer.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsPositiveModelScorer.name</code> (String?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsPositiveModelScorer.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsTrainingScores.name</code> (String?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsTrainingScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsUnlabeledPositiveModelScores.name</code> (String?)</li>
<li><code>FinalizeSnpTrainVariantAnnotationsUnlabeledPositiveModelScores.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeVETSTBI.name</code> (String?)</li>
<li><code>FinalizeVETSTBI.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeVETSVCF.name</code> (String?)</li>
<li><code>FinalizeVETSVCF.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FinalizeZarr.name</code> (String?)</li>
<li><code>FinalizeZarr.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>FunctionallyAnnotate.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>GatherRescoredVcfs.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>GnarlyJointCallGVCFs.input_gvcf_index</code> (File?)</li>
<li><code>JointCallGVCFs.input_gvcf_index</code> (File?)</li>
<li><code>MakeIntervalListFromSequenceDictionary.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>MakeSitesOnlyVCF.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>MergeSitesOnlyVCFs.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ScoreIndelVariantAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>ScoreSnpVariantAnnotations.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>TrainIndelVariantAnnotationsModel.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>TrainIndelVariantAnnotationsModel.unlabeled_annotation_hdf5</code> (File?)</li>
<li><code>TrainSnpVariantAnnotationsModel.runtime_attr_override</code> (RuntimeAttr?)</li>
<li><code>TrainSnpVariantAnnotationsModel.unlabeled_annotation_hdf5</code> (File?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>do_zarr_conversion</code> (Boolean, default=false)</li>
<li><code>heterozygosity</code> (Float, default=0.001): Joint Genotyping Parameter - Heterozygosity value used to compute prior likelihoods for any locus. See the GATKDocs for full details on the meaning of this population genetics concept</li>
<li><code>heterozygosity_stdev</code> (Float, default=0.01): Joint Genotyping Parameter - Standard deviation of heterozygosity for SNP and indel calling.</li>
<li><code>indel_calibration_sensitivity</code> (Float, default=0.99): VETS (ScoreVariantAnnotations) parameter - score below which INDEL variants will be filtered.</li>
<li><code>indel_heterozygosity</code> (Float, default=0.000125): Joint Genotyping Parameter - Heterozygosity for indel calling. See the GATKDocs for heterozygosity for full details on the meaning of this population genetics concept</li>
<li><code>indel_max_unlabeled_variants</code> (Int, default=0): VETS (ExtractVariantAnnotations) parameter - maximum number of unlabeled INDEL variants/alleles to randomly sample with reservoir sampling.  If nonzero, annotations will also be extracted from unlabeled sites.</li>
<li><code>indel_recalibration_annotation_values</code> (Array[String], default=["BaseQRankSum", "ExcessHet", "FS", "MQ", "MQRankSum", "QD", "ReadPosRankSum", "SOR", "DP"]): VETS (ScoreSnpVariantAnnotations/ScoreVariantAnnotations) parameter - Array of annotation names to use to create the INDEL variant scoring model and over which to score INDEL variants.</li>
<li><code>shard_max_interval_size_bp</code> (Int, default=999999999): Maximum size of the interval on each shard.  This along with the given sequence dictionary determines how many shards there will be.  To shard by contig, set to a very high number.  Default is 999999999.</li>
<li><code>snp_calibration_sensitivity</code> (Float, default=0.99): VETS (ScoreVariantAnnotations) parameter - score below which SNP variants will be filtered.</li>
<li><code>snp_max_unlabeled_variants</code> (Int, default=0): VETS (ExtractVariantAnnotations) parameter - maximum number of unlabeled SNP variants/alleles to randomly sample with reservoir sampling.  If nonzero, annotations will also be extracted from unlabeled sites.</li>
<li><code>snp_recalibration_annotation_values</code> (Array[String], default=["BaseQRankSum", "ExcessHet", "FS", "MQ", "MQRankSum", "QD", "ReadPosRankSum", "SOR", "DP"]): VETS (ScoreSnpVariantAnnotations/ScoreVariantAnnotations) parameter - Array of annotation names to use to create the SNP variant scoring model and over which to score SNP variants.</li>
<li><code>use_gnarly_genotyper</code> (Boolean, default=false): If true, the <code>GnarlyGenotyper</code> will be used, greatly speeding up joint genotyping (at the cost of potentially lower accuracy).  Setting this to true is recommended for large callsets.  If false, <code>GenotypeGVCFs</code> will be used to generate the final VCF.  Default is false.</li>
<li><code>ConvertToZarr.num_cpus</code> (Int, default=4)</li>
<li><code>ConvertToZarr.reference</code> (String, default="GRCh38")</li>
<li><code>GnarlyJointCallGVCFs.keep_combined_raw_annotations</code> (Boolean, default=false)</li>
<li><code>ImportGVCFsIntoGenomicsDB.extra_mem_gb</code> (Int, default=0)</li>
<li><code>JointCallGVCFs.keep_combined_raw_annotations</code> (Boolean, default=false)</li>
<li><code>MakeIntervalListFromSequenceDictionary.ignore_contigs</code> (Array[String], default=['random', 'chrUn', 'decoy', 'alt', 'HLA', 'EBV'])</li>
<li><code>TrainIndelVariantAnnotationsModel.calibration_sensitivity_threshold</code> (Float, default=0.95)</li>
<li><code>TrainSnpVariantAnnotationsModel.calibration_sensitivity_threshold</code> (Float, default=0.95)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>joint_vcf_out</code> (File)</li>
<li><code>joint_vcf_out_tbi</code> (File)</li>
<li><code>joint_mt</code> (File)</li>
<li><code>genomicsDB</code> (Array[String])</li>
<li><code>joint_zarr</code> (File?)</li>
<li><code>snpEff_summary</code> (Array[String]?)</li>
<li><code>snpEff_genes</code> (Array[String]?)</li>
</ul>
<h3 id="dot-diagram">Dot Diagram</h3>
<p><img alt="SRJointCallGVCFsWithGenomicsDB" src="../../dot/SRJointCallGVCFsWithGenomicsDB.dot.png" /></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../SRIndexBam/" class="btn btn-neutral float-left" title="SRIndexBam"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS/" class="btn btn-neutral float-right" title="SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright (c) 2023, Broad Institute</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/broadinstitute/long-read-pipelines" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../SRIndexBam/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
