<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Broad Institute" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>LRCNVs - Broad Institute Long Read Pipelines</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "LRCNVs";
        var mkdocs_page_input_path = "workflows/LRCNVs.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> Broad Institute Long Read Pipelines
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Development guide</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/">Development Guide Overview</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/docker_style_guide/">Docker Style Guide</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/repo_structure/">Repository Structure</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../development_guide/wdl_style_guide/">WDL Style Guide</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Pipeline documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="#">Sp malaria</a>
    <ul>
                <li class="toctree-l2"><a class="reference internal" href="../../pipeline_documentation/sp_malaria/workspace_markdown_doc/">_P. falciparum_ Short Read Whole Genome Workspace</a>
                </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/AlignReads/">AlignReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/BroadOnPremMalariaPipelineTasks/">BroadOnPremMalariaPipelineTasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Clair/">Clair</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/FastQC/">FastQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Finalize/">Finalize</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Fingerprinting/">Fingerprinting</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/FunctionalAnnotation/">FunctionalAnnotation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/GeneralUtils/">GeneralUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Hail/">Hail</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MASSeq/">MASSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MalariaReports/">MalariaReports</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/MalariaReportsSnapshots/">MalariaReportsSnapshots</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Medaka/">Medaka</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/NanoPlot/">NanoPlot</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ONTPepper/">ONTPepper</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ONTUtils/">ONTUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/PBUtils/">PBUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Postprocessing_Tasks/">Postprocessing_Tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Preprocessing_Tasks/">Preprocessing_Tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/QCAssessment/">QCAssessment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Quast/">Quast</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SGKit/">SGKit</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SRJointGenotyping/">SRJointGenotyping</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/SRUtils/">SRUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/StringTie2/">StringTie2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/Utils/">Utils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/VariantUtils/">VariantUtils</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cnv_common_tasks/">cnv_common_tasks</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../AlignAndCheckFingerprintCCS/">AlignAndCheckFingerprintCCS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AlignedMetrics/">AlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AnnotateTranscriptomeWithGuide/">AnnotateTranscriptomeWithGuide</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../AssignChildLongReads/">AssignChildLongReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BackupWorkspace/">BackupWorkspace</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BenchmarkVCFs/">BenchmarkVCFs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BroadOnPremMalariaPipeline_1_Alignment_and_Variant_Calling/">BroadOnPremMalariaPipeline_1_Alignment_and_Variant_Calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../BroadOnPremMalariaPipeline_2_JointVariantCalling/">BroadOnPremMalariaPipeline_2_JointVariantCalling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CCSPepper/">CCSPepper</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallAssemblyVariants/">CallAssemblyVariants</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsIllumina/">CallVariantsIllumina</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsONT/">CallVariantsONT</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CallVariantsPBCCS/">CallVariantsPBCCS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Canu/">Canu</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CleanupIntermediate/">CleanupIntermediate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectPacBioAlignedMetrics/">CollectPacBioAlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectParentsKmerStats/">CollectParentsKmerStats</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CollectSMRTCellUnalignedMetrics/">CollectSMRTCellUnalignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../CompareVcfBenchmarks/">CompareVcfBenchmarks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ConvertToHailMT/">ConvertToHailMT</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ConvertToZarrStore/">ConvertToZarrStore</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DeepTrio/">DeepTrio</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DeepVariant/">DeepVariant</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromFTP/">DownloadFromFTP</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromSRA/">DownloadFromSRA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../DownloadFromWeb/">DownloadFromWeb</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExpandedDrugResistanceMarkerAggregation/">ExpandedDrugResistanceMarkerAggregation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExpandedDrugResistanceMarkerExtraction/">ExpandedDrugResistanceMarkerExtraction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ExtractRegionsFromBam/">ExtractRegionsFromBam</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../FPCheckAoU/">FPCheckAoU</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../FindBamIdentity/">FindBamIdentity</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Flye/">Flye</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../GLNexus/">GLNexus</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Guppy/">Guppy</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../HaplotypeCaller/">HaplotypeCaller</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Hifiasm/">Hifiasm</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">LRCNVs</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#lrcnvs_1">LRCNVs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#dot-diagram">Dot Diagram</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LRConvertBCF/">LRConvertBCF</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../LRJointCallGVCFs/">LRJointCallGVCFs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Longbow/">Longbow</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../MalariaReportsWorkflow/">MalariaReportsWorkflow</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTAssembleWithCanu/">ONTAssembleWithCanu</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTAssembleWithFlye/">ONTAssembleWithFlye</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTBasecall/">ONTBasecall</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTFlowcell/">ONTFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTFlowcellFromMultipleBasecalls/">ONTFlowcellFromMultipleBasecalls</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTMethylation/">ONTMethylation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTProcessBasecall/">ONTProcessBasecall</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ONTWholeGenome/">ONTWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBAssembleWithHifiasm/">PBAssembleWithHifiasm</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBCCSIsoSeq/">PBCCSIsoSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBCCSWholeGenome/">PBCCSWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBFlowcell/">PBFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBMASIsoSeqDemultiplex/">PBMASIsoSeqDemultiplex</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBMASIsoSeqQuantify/">PBMASIsoSeqQuantify</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PBSV/">PBSV</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PanelProcessMalariaBarcodesForRh/">PanelProcessMalariaBarcodesForRh</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Pf_Niare_HaplotypeCaller/">Pf_Niare_HaplotypeCaller</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumDrugResistanceSummary/">PfalciparumDrugResistanceSummary</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumHrp2Hrp3Status/">PfalciparumHrp2Hrp3Status</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumPolygenomicityBarcodeEstimate/">PfalciparumPolygenomicityBarcodeEstimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../PfalciparumTypeDrugResistanceMarkers/">PfalciparumTypeDrugResistanceMarkers</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ProcessMalariaBarcodesDemo/">ProcessMalariaBarcodesDemo</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../RemoveSingleOrganismContamination/">RemoveSingleOrganismContamination</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRBamToFq/">SRBamToFq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRFlowcell/">SRFlowcell</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRIndexBam/">SRIndexBam</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRJointCallGVCFsWithGenomicsDB/">SRJointCallGVCFsWithGenomicsDB</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS/">SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VETS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VQSR/">SRJointCallGVCFsWithGenomicsDB_Pf_Niare_VQSR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome/">SRWholeGenome</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome_Pf_Niare_VETS/">SRWholeGenome_Pf_Niare_VETS</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SRWholeGenome_Pf_Niare_VQSR/">SRWholeGenome_Pf_Niare_VQSR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SampleLevelAlignedMetrics/">SampleLevelAlignedMetrics</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../Sniffles2/">Sniffles2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../SvQCPlots/">SvQCPlots</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TrainCnnFilters/">TrainCnnFilters</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../TrioBinChildLongReads/">TrioBinChildLongReads</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../VerifyFingerprint/">VerifyFingerprint</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">Broad Institute Long Read Pipelines</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Workflows &raquo;</li>
      <li>LRCNVs</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/broadinstitute/long-read-pipelines/edit/master/docs/workflows/LRCNVs.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="lrcnvs">LRCNVs</h1>
<h2 id="lrcnvs_1">LRCNVs</h2>
<dl>
<dt>description</dt>
<dd>Workflow for creating a GATK GermlineCNVCaller denoising model and generating calls given a list of normal samples. Supports both WGS and WES.  The intervals argument is required for both WGS and WES workflows and accepts formats compatible with the GATK -L argument (see https://gatkforums.broadinstitute.org/gatk/discussion/11009/intervals-and-interval-lists). These intervals will be padded on both sides by the amount specified by padding (default 250) and split into bins of length specified by bin_length (default 1000; specify 0 to skip binning, e.g., for WES).  For WGS, the intervals should simply cover the chromosomes of interest.</dd>
<dt>notes</dt>
<dd><br /> - The intervals argument is required for both WGS and WES workflows and accepts formats compatible with the GATK -L argument (see https://gatkforums.broadinstitute.org/gatk/discussion/11009/intervals-and-interval-lists). These intervals will be padded on both sides by the amount specified by padding (default 250) and split into bins of length specified by bin_length (default 1000; specify 0 to skip binning, e.g., for WES).  For WGS, the intervals should simply cover the chromosomes of interest. <br />  - Intervals can be blacklisted from coverage collection and all downstream steps by using the blacklist_intervals argument, which accepts formats compatible with the GATK -XL argument (see https://gatkforums.broadinstitute.org/gatk/discussion/11009/intervals-and-interval-lists). This may be useful for excluding centromeric regions, etc. from analysis.  Alternatively, these regions may be manually filtered from the final callset.</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>cohort_entity_id</code> (String, <strong>required</strong>): Required.  String identifier for the cohort used for denoising model generation.</li>
<li><code>contig_ploidy_priors</code> (File, <strong>required</strong>): Required.  File containing contig ploidy priors.</li>
<li><code>intervals</code> (File, <strong>required</strong>): Required.  Intervals file (in .bed format) containing the regions to be analyzed.  For WGS, this should simply cover the chromosomes of interest.  For WES, this should cover the target regions to be analyzed.</li>
<li><code>normal_bais</code> (Array[String]+, <strong>required</strong>): Required.  Array of BAI index files for the normal samples to be used for denoising model generation.</li>
<li><code>normal_bams</code> (Array[String]+, <strong>required</strong>): Required.  Array of BAM files for the normal samples to be used for denoising model generation.</li>
<li><code>ref_map_file</code> (File, <strong>required</strong>): Required.  File containing the reference copy-number map.</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>allosomal_contigs</code> (Array[String]?)</li>
<li><code>bin_length</code> (Int?)</li>
<li><code>blacklist_intervals</code> (File?): Optional.  Intervals file (in .bed format) containing the regions to be excluded from analysis.  For WGS, this may be useful for excluding centromeric regions, etc. from analysis.  For WES, this may be useful for excluding regions that are not covered by the target capture kit.</li>
<li><code>blacklist_intervals_for_filter_intervals</code> (File?)</li>
<li><code>collect_counts_enable_indexing</code> (Boolean?)</li>
<li><code>collect_counts_format</code> (String?)</li>
<li><code>cpu_for_determine_germline_contig_ploidy</code> (Int?)</li>
<li><code>cpu_for_germline_cnv_caller</code> (Int?)</li>
<li><code>do_explicit_gc_correction</code> (Boolean?)</li>
<li><code>extreme_count_filter_maximum_percentile</code> (Float?)</li>
<li><code>extreme_count_filter_minimum_percentile</code> (Float?)</li>
<li><code>extreme_count_filter_percentage_of_samples</code> (Float?)</li>
<li><code>feature_query_lookahead</code> (Int?)</li>
<li><code>gatk4_jar_override</code> (File?)</li>
<li><code>gcnv_active_class_padding_hybrid_mode</code> (Int?)</li>
<li><code>gcnv_adamax_beta_1</code> (Float?)</li>
<li><code>gcnv_adamax_beta_2</code> (Float?)</li>
<li><code>gcnv_caller_external_admixing_rate</code> (Float?)</li>
<li><code>gcnv_caller_internal_admixing_rate</code> (Float?)</li>
<li><code>gcnv_caller_update_convergence_threshold</code> (Float?)</li>
<li><code>gcnv_class_coherence_length</code> (Float?)</li>
<li><code>gcnv_cnv_coherence_length</code> (Float?)</li>
<li><code>gcnv_convergence_snr_averaging_window</code> (Int?)</li>
<li><code>gcnv_convergence_snr_countdown_window</code> (Int?)</li>
<li><code>gcnv_convergence_snr_trigger_threshold</code> (Float?)</li>
<li><code>gcnv_copy_number_posterior_expectation_mode</code> (String?)</li>
<li><code>gcnv_depth_correction_tau</code> (Float?)</li>
<li><code>gcnv_disable_annealing</code> (Boolean?)</li>
<li><code>gcnv_enable_bias_factors</code> (Boolean?)</li>
<li><code>gcnv_gc_curve_standard_deviation</code> (Float?)</li>
<li><code>gcnv_init_ard_rel_unexplained_variance</code> (Float?)</li>
<li><code>gcnv_initial_temperature</code> (Float?)</li>
<li><code>gcnv_interval_psi_scale</code> (Float?)</li>
<li><code>gcnv_learning_rate</code> (Float?)</li>
<li><code>gcnv_log_emission_samples_per_round</code> (Int?)</li>
<li><code>gcnv_log_emission_sampling_median_rel_error</code> (Float?)</li>
<li><code>gcnv_log_emission_sampling_rounds</code> (Int?)</li>
<li><code>gcnv_log_mean_bias_standard_deviation</code> (Float?)</li>
<li><code>gcnv_mapping_error_rate</code> (Float?)</li>
<li><code>gcnv_max_advi_iter_first_epoch</code> (Int?)</li>
<li><code>gcnv_max_advi_iter_subsequent_epochs</code> (Int?)</li>
<li><code>gcnv_max_bias_factors</code> (Int?)</li>
<li><code>gcnv_max_calling_iters</code> (Int?)</li>
<li><code>gcnv_max_copy_number</code> (Int?)</li>
<li><code>gcnv_max_training_epochs</code> (Int?)</li>
<li><code>gcnv_min_training_epochs</code> (Int?)</li>
<li><code>gcnv_num_gc_bins</code> (Int?)</li>
<li><code>gcnv_num_thermal_advi_iters</code> (Int?)</li>
<li><code>gcnv_p_active</code> (Float?)</li>
<li><code>gcnv_p_alt</code> (Float?)</li>
<li><code>gcnv_sample_psi_scale</code> (Float?)</li>
<li><code>low_count_filter_count_threshold</code> (Int?)</li>
<li><code>low_count_filter_percentage_of_samples</code> (Float?)</li>
<li><code>mappability_track_bed</code> (File?)</li>
<li><code>mappability_track_bed_idx</code> (File?)</li>
<li><code>maximum_gc_content</code> (Float?)</li>
<li><code>maximum_mappability</code> (Float?)</li>
<li><code>maximum_segmental_duplication_content</code> (Float?)</li>
<li><code>mem_gb_for_annotate_intervals</code> (Int?)</li>
<li><code>mem_gb_for_collect_counts</code> (Int?)</li>
<li><code>mem_gb_for_determine_germline_contig_ploidy</code> (Int?)</li>
<li><code>mem_gb_for_filter_intervals</code> (Int?)</li>
<li><code>mem_gb_for_germline_cnv_caller</code> (Int?)</li>
<li><code>minimum_gc_content</code> (Float?)</li>
<li><code>minimum_mappability</code> (Float?)</li>
<li><code>minimum_segmental_duplication_content</code> (Float?)</li>
<li><code>padding</code> (Int?)</li>
<li><code>ploidy_global_psi_scale</code> (Float?)</li>
<li><code>ploidy_mapping_error_rate</code> (Float?)</li>
<li><code>ploidy_mean_bias_standard_deviation</code> (Float?)</li>
<li><code>ploidy_sample_psi_scale</code> (Float?)</li>
<li><code>preemptible_attempts</code> (Int?)</li>
<li><code>segmental_duplication_track_bed</code> (File?)</li>
<li><code>segmental_duplication_track_bed_idx</code> (File?)</li>
<li><code>AnnotateIntervals.cpu</code> (Int?)</li>
<li><code>AnnotateIntervals.disk_space_gb</code> (Int?)</li>
<li><code>CollectCounts.cpu</code> (Int?)</li>
<li><code>CollectCounts.disk_space_gb</code> (Int?)</li>
<li><code>CollectModelQualityMetrics.cpu</code> (Int?)</li>
<li><code>CollectModelQualityMetrics.disk_space_gb</code> (Int?)</li>
<li><code>CollectModelQualityMetrics.mem_gb</code> (Int?)</li>
<li><code>CollectSampleQualityMetrics.cpu</code> (Int?)</li>
<li><code>CollectSampleQualityMetrics.disk_space_gb</code> (Int?)</li>
<li><code>CollectSampleQualityMetrics.mem_gb</code> (Int?)</li>
<li><code>DetermineGermlineContigPloidyCohortMode.disk_space_gb</code> (Int?)</li>
<li><code>DetermineGermlineContigPloidyCohortMode.output_dir</code> (String?)</li>
<li><code>FilterIntervals.cpu</code> (Int?)</li>
<li><code>FilterIntervals.disk_space_gb</code> (Int?)</li>
<li><code>GermlineCNVCallerCohortMode.disk_space_gb</code> (Int?)</li>
<li><code>GermlineCNVCallerCohortMode.output_dir</code> (String?)</li>
<li><code>PostprocessGermlineCNVCalls.cpu</code> (Int?)</li>
<li><code>PostprocessGermlineCNVCalls.disk_space_gb</code> (Int?)</li>
<li><code>PostprocessGermlineCNVCalls.mem_gb</code> (Int?)</li>
<li><code>PreprocessIntervals.cpu</code> (Int?)</li>
<li><code>PreprocessIntervals.disk_space_gb</code> (Int?)</li>
<li><code>PreprocessIntervals.mem_gb</code> (Int?)</li>
<li><code>ScatterIntervals.cpu</code> (Int?)</li>
<li><code>ScatterIntervals.disk_space_gb</code> (Int?)</li>
<li><code>ScatterIntervals.gatk4_jar_override</code> (File?)</li>
<li><code>ScatterIntervals.mem_gb</code> (Int?)</li>
<li><code>ScatterIntervals.output_dir</code> (String?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>gatk_docker</code> (String, default="broadinstitute/gatk:4.1.7.0"): Optional.  Docker image for the GATK tool.  Default is broadinstitute/gatk:broadinstitute/gatk:4.1.7.0.</li>
<li><code>maximum_number_events_per_sample</code> (Int, default=1000)</li>
<li><code>num_intervals_per_scatter</code> (Int, default=10000): Optional.  Number of intervals to process in each scatter.  Default is 10000.</li>
<li><code>ref_copy_number_autosomal_contigs</code> (Int, default=2)</li>
<li><code>AnnotateIntervals.use_ssd</code> (Boolean, default=false)</li>
<li><code>CollectCounts.use_ssd</code> (Boolean, default=false)</li>
<li><code>CollectModelQualityMetrics.use_ssd</code> (Boolean, default=false)</li>
<li><code>CollectSampleQualityMetrics.use_ssd</code> (Boolean, default=false)</li>
<li><code>DetermineGermlineContigPloidyCohortMode.use_ssd</code> (Boolean, default=false)</li>
<li><code>FilterIntervals.use_ssd</code> (Boolean, default=false)</li>
<li><code>GermlineCNVCallerCohortMode.use_ssd</code> (Boolean, default=false)</li>
<li><code>PostprocessGermlineCNVCalls.use_ssd</code> (Boolean, default=false)</li>
<li><code>PreprocessIntervals.use_ssd</code> (Boolean, default=false)</li>
<li><code>ScatterIntervals.use_ssd</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>preprocessed_intervals</code> (File)</li>
<li><code>read_counts_entity_ids</code> (Array[File])</li>
<li><code>read_counts</code> (Array[File])</li>
<li><code>annotated_intervals</code> (File?)</li>
<li><code>filtered_intervals</code> (File)</li>
<li><code>contig_ploidy_model_tar</code> (File)</li>
<li><code>contig_ploidy_calls_tar</code> (File)</li>
<li><code>gcnv_model_tars</code> (Array[File])</li>
<li><code>gcnv_calls_tars</code> (Array[Array[File]])</li>
<li><code>gcnv_tracking_tars</code> (Array[File])</li>
<li><code>genotyped_intervals_vcfs</code> (Array[File])</li>
<li><code>genotyped_segments_vcfs</code> (Array[File])</li>
<li><code>sample_qc_status_files</code> (Array[File])</li>
<li><code>sample_qc_status_strings</code> (Array[String])</li>
<li><code>model_qc_status_file</code> (File)</li>
<li><code>model_qc_string</code> (String)</li>
<li><code>denoised_copy_ratios</code> (Array[File])</li>
</ul>
<h3 id="dot-diagram">Dot Diagram</h3>
<p><img alt="LRCNVs" src="../../dot/LRCNVs.dot.png" /></p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../Hifiasm/" class="btn btn-neutral float-left" title="Hifiasm"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../LRConvertBCF/" class="btn btn-neutral float-right" title="LRConvertBCF">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright (c) 2023, Broad Institute</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/broadinstitute/long-read-pipelines" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../Hifiasm/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../LRConvertBCF/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
